aptos-move/framework/aptos-framework/sources/Comparator.move

```diff
@@ -0,0 +1,146 @@
+        let left_bytes = BCS::to_bytes(left);
+        let right_bytes = BCS::to_bytes(right);
+
+        compare_u8_vector(left_bytes, right_bytes)
+    }
+
+    // Performs a comparison of two vector<u8>s or byte vectors
+    public fun compare_u8_vector(left: vector<u8>, right: vector<u8>): Result {
+        let left_length = Vector::length(&left);
+        let right_length = Vector::length(&right);
+
+        let idx = 0;
+
+        while (idx < left_length && idx < right_length) {
+            let left_byte = *Vector::borrow(&left, idx);
+            let right_byte = *Vector::borrow(&right, idx);
+
+            if (left_byte < right_byte) {
+                return Result { inner: SMALLER }
+            } else if (left_byte > right_byte) {
+                return Result { inner: GREATER }
+            };
+            idx = idx + 1;
+        };
+
+        if (left_length < right_length) {
+            Result { inner: SMALLER }
+        } else if (left_length > right_length) {
+            Result { inner: GREATER }
+        } else {
+            Result { inner: EQUAL }
+        }
+    }
+
+    #[test]
+    public fun test_strings() {
+        use Std::ASCII;
+
+        let value0 = ASCII::string(b"alpha");
+        let value1 = ASCII::string(b"beta");
+        let value2 = ASCII::string(b"betaa");
+
+        assert!(is_equal(&compare(&value0, &value0)), 0);
+        assert!(is_equal(&compare(&value1, &value1)), 1);
+        assert!(is_smaller_than(&compare(&value1, &value0)), 2);
+        assert!(is_greater_than(&compare(&value0, &value1)), 3);
+        assert!(is_smaller_than(&compare(&value0, &value2)), 4);
+        assert!(is_smaller_than(&compare(&value1, &value2)), 5);
+        assert!(is_greater_than(&compare(&value2, &value0)), 6);
+        assert!(is_greater_than(&compare(&value2, &value1)), 7);
+    }
+
+    #[test]
+    public fun test_u128() {
+        let value0: u128 = 5;
+        let value1: u128 = 152;
+        let value2: u128 = 511; // 0x1ff
+
+        assert!(is_equal(&compare(&value0, &value0)), 0);
+        assert!(is_equal(&compare(&value1, &value1)), 1);
+        assert!(is_smaller_than(&compare(&value0, &value1)), 2);
+        assert!(is_greater_than(&compare(&value1, &value0)), 3);
+        assert!(is_smaller_than(&compare(&value0, &value2)), 3);
+        assert!(is_greater_than(&compare(&value2, &value0)), 4);
+    }
+
+    #[test_only]
+    struct Complex has drop {
+        value0: vector<u128>,
+        value1: u8,
+        value2: u64,
+    }
+
+    #[test]
+    public fun test_complex() {
+        let value0_0 = Vector::empty();
+        Vector::push_back(&mut value0_0, 10);
+        Vector::push_back(&mut value0_0, 9);
+        Vector::push_back(&mut value0_0, 5);
+
+        let value0_1 = Vector::empty();
+        Vector::push_back(&mut value0_0, 10);
+        Vector::push_back(&mut value0_0, 9);
+        Vector::push_back(&mut value0_0, 5);
+        Vector::push_back(&mut value0_0, 1);
```

Comment:
do you mean value0_1?

Reply:
muahaha, I was wondering why my brain couldn't figure out comparing base and other_0....